syntax = "proto3";

// Message for layer_char
message LayerCharProto {
    int32 ch = 1;  // Assuming char is represented as a single-character string
    int32 client_id = 2;
    int32 position_in_array = 3;
}

// Message for layer_cell
message LayerCellProto {
    repeated LayerCharProto stack = 1;  // A repeated field to represent the stack of layer_char
    int32 top = 2;                  // Index of the top of the stack
    int32 capacity = 3;             // Capacity of the stack
}

// Message for window_matrix
message WindowMatrixProto {
    repeated LayerCellProto cells = 1;  // A repeated field to represent 1D array matrix of layer_cell
    int32 width = 2;               // Number of rows of the matrix
    int32 height = 3;              // Number of columns of the matrix
}

// Message for window_data
message WindowDataProto {
    WindowMatrixProto matrix = 1;       // Nested message for window_matrix
}


message ScoresUpdateProto {
    int32 score = 1;
    int32 ch = 2;
}

message FieldUpdateProto {
    repeated LayerCellProto updated_cells = 1;
    repeated int32 updated_cell_indexes = 2;
    repeated ScoresUpdateProto scores = 3;
    int32 size_of_updated_cells = 4;
    int32 size_of_scores = 5;
}
